aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -0,0 +1,1211 @@
+                    self.txn_status[dep_txn_idx as usize]
+                        .resolve_dependency(DependencyInstruction::WaitForExecuting)?
+                },
+                (Low, Some(true)) | (Medium, Some(true)) => {
+                    let ret = if matches!(priority, Medium) {
+                        self.txn_status[dep_txn_idx as usize]
+                            .resolve_dependency(DependencyInstruction::WaitForExecuting)?
+                    } else {
+                        DependencyResolution::None
+                    };
+
+                    // The logic of scheduling first re-executions after all prior transactions
+                    // have finished 0-th incarnations is useful if the writes of these 0-th
+                    // incarnations are properly recorded and available for reading. Hence, we
+                    // disallow speculatively aborting 0-th incarnation. An alternative would be
+                    // to only allow increasing scheduler's executed_idx after an execution that
+                    // is not speculatively aborted (such as here).
+                    if incarnation > 0 && matches!(ret, DependencyResolution::None) {
+                        let to_propagate = self.try_abort(txn_idx, incarnation)?;
+                        self.propagate(to_propagate)?;
+
+                        // The current logic is to just ask caller to speculatively abort.
+                        // TODO: complex handling (e.g. scheduler maintenance), also revisit Wait.
+                        return Ok(false);
+                    }
+
+                    ret
+                },
+                (Low, Some(false)) | (Medium, Some(false)) => {
+                    // There is value in finishing the speculative execution, as the transaction
+                    // has not been previously aborted due to this dependency: proceeding allows
+                    // an optimistic approach, and recording the aborted dependency upon failure.
+                    DependencyResolution::None
+                },
+                (High | Medium | Low, None) | (Highest, Some(_)) => {
+                    unreachable!("abort_likely checked for high / medium / low priority")
+                },
+            };
+
+            match resolution {
+                DependencyResolution::SafeToProceed | DependencyResolution::None => {
+                    return Ok(true);
+                },
+                DependencyResolution::Wait(dep_condition) => {
+                    Self::wait(dep_condition);
+                    // Next iteration of the loop will try to resolve again.
+                },
+                DependencyResolution::Halted => {
+                    return Ok(false);
+                },
+            }
+        }
+    }
+
+    pub(crate) fn finish_execution(
+        &self,
+        txn_idx: TxnIndex,
+        incarnation: Incarnation,
+        invalidated_versions: BTreeSet<(TxnIndex, Incarnation)>,
```

Comment:
not very clear where this comes from? this is for later version that's invalidated by this txn's writes?

Reply:
yes, should I add a comment or any other way I can make it more clear?

Reply:
This is now taken care of, forces to pass a struct by value. That struct is the proper way to manage invalidations / aborts during an execution (and maintains internal state, which we take inside finish_execution)